util/stderr_logger.cc

```diff
@@ -19,12 +19,36 @@ void StderrLogger::Logv(const char* format, va_list ap) {
   const time_t seconds = now_tv.tv_sec;
   struct tm t;
   port::LocalTimeR(&seconds, &t);
-  fprintf(stderr, "%04d/%02d/%02d-%02d:%02d:%02d.%06d %llx ", t.tm_year + 1900,
-          t.tm_mon + 1, t.tm_mday, t.tm_hour, t.tm_min, t.tm_sec,
-          static_cast<int>(now_tv.tv_usec),
-          static_cast<long long unsigned int>(thread_id));
 
-  vfprintf(stderr, format, ap);
-  fprintf(stderr, "\n");
+  // The string we eventually log has three parts: the context (time, thread),
+  // optional user-supplied prefix, and the actual log message.
+  //
+  // We compute their lengths so that we can allocate a buffer big enough to
+  // print it. The context string (with the date and thread id) is really only
+  // 36 bytes, but we allocate 40 to be safe.
+  size_t context_len = 40;
+
+  va_list ap_copy;
+  va_copy(ap_copy, ap);
+  const size_t log_suffix_len = vsnprintf(nullptr, 0, format, ap_copy);
+  va_end(ap_copy);
+
+  // Allocate space for the context, log_prefix, and log itself
+  // Extra byte for null termination
+  size_t buf_len = context_len + log_prefix_len + log_suffix_len + 1;
+  std::unique_ptr<char[]> buf(new char[buf_len]);
+
+  // Write out the context and prefix string
+  int written =
+      snprintf(buf.get(), context_len + log_prefix_len,
+               // Keep this string in sync with context_len
+               "%04d/%02d/%02d-%02d:%02d:%02d.%06d %llx %s", t.tm_year + 1900,
```

Comment:
Why is context_len 40, does thread_id not take 16 chars?

Reply:
The `thread_id` gets cast into a `long long unsigned int`. To be explicit, I got 40 because:

`int/int/int-int:int:int.int longlong` is 4 + 4 + 4 + 4 + 4 + 4 + 4 + 8 = 36.

Am I missing something?

Reply:
I got:
```
%04d/%02d/%02d-%02d:%02d:%02d.%06d %llx %s
4 + 1+2+ 1+2 +1+2+ 1+2+ 1+2+ 1+ 6 + 16+1  = 43
```

Reply:
Yeah, this is closer—I forgot to count lengths. But actually, I think the 43 is one too little, I think you forgot the space before the `%llx`. So I think it is actually 44. Since this seems to be very error prone, I have included a comment with counting in the source.